README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# FishMap <img src="man/figures/logo.png" align="right" alt="" width="120" />

<!-- badges: start -->

[![.github/workflows/R-CMD-check](https://github.com/balglave/FishMap/actions/workflows/.github/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/balglave/FishMap/actions/workflows/.github/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/balglave/FishMap/branch/main/graph/badge.svg)](https://app.codecov.io/gh/balglave/FishMap?branch=main)
<!-- badges: end -->

The goal of FishMap is to …

## Installation

### Dependencies

You must [install INLA](https://www.r-inla.org/download-install).
Checkout [INLA doc](https://www.r-inla.org/download-install) to upgrade
INLA. Your INLA version should satisfy the minimal required version for
running FishMap (22.12.16).

``` r
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(c("graph", "Rgraphviz"), dep=TRUE)

install.packages("INLA",repos=c(getOption("repos"),INLA="https://inla.r-inla-download.org/R/stable"), dep=TRUE)
```

You must [install TMB](https://github.com/kaskr/adcomp/wiki/Download).
The installation of TMB requires compiler tools. If you are installing
from Windows machine, make sure that your [Rtools installation is
matching your
R-version](https://github.com/kaskr/adcomp/wiki/Download#windows---installing-required-tools).
Your TMB version should satisfy the minimal required version for running
FishMap (1.9.2).

``` r
install.packages("TMB")
```

### FishMap

You can install the development version of FishMap from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("balglave/FishMap")
```

### Vignettes

You can install the development version of FishMap together with the
vignettes with:

``` r
# install.packages("remotes")
remotes::install_github("balglave/FishMap", build_vignettes = TRUE)
```

Installing with `build_vignettes = TRUE` will allow you to access the
vignettes locally. Once you have installed FishMap with vignettes, you
can access them using the command `vignette(package = "FishMap")`.

## Documentation

Full documentation website on: <https://balglave.github.io/FishMap>  
In particular, you can follow this vignette:

``` r
vignette("user-running-fishmap", package = "FishMap")
```

#### Unit tests in the package :

Writing unit tests is an essential practice in software development that
can greatly improve the quality and maintainability of your code. Here
are some of the key reasons why writing unit tests is important:

-   Ensuring correctness: Unit tests help ensure that your code works as
    intended and meets the requirements of the system. By writing tests
    that cover a wide range of scenarios and edge cases, you can catch
    bugs and errors before they become bigger problems down the line.

-   Facilitating code changes: Unit tests can also help you make changes
    to your code with confidence. By having a comprehensive suite of
    tests that can quickly validate changes, you can refactor and
    optimize your codebase without fear of introducing new bugs or
    breaking existing functionality.

-   Saving time and money: Catching bugs early in the development
    process can save time and money in the long run. By reducing the
    time spent on debugging and fixing issues, you can focus more on
    delivering new features and improvements to your application.

-   Promoting code quality: Writing tests can also encourage good coding
    practices, such as writing modular and reusable code. By breaking
    down your code into smaller, testable units, you can make your code
    more flexible and maintainable over time.

Overall, unit testing is a critical practice that can improve the
quality, maintainability, and reliability of your software. By investing
time and effort into writing good tests, you can ensure that your code
works as intended and can adapt to changing requirements and user needs.

You’ll find under this
[link](https://balglave.github.io/FishMap/testdown/testdown-report-for-fishmap.html),
the testdown report that describe the tests implemented for this package
and you will find in this
[vignette](https://balglave.github.io/FishMap/articles/dev_how_to_maintain_my_tests_and_examples.html)
the technical choices made for the realization of these tests.

## Workflow

This workflow is intended for developers, it summaries the structure of
the package. For each flat file in the `dev/` folder, you have a set of
defined functions as well as the list of files generated from the flat
file when running the `fusen::inflate()` command.

<img src="man/figures/FishMap_user.png" width="100%" />

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(FishMap)

# Read internal data as Rds files
survey_data_file <- system.file("original_data",
                                "Solea_solea",
                                "survey_data.Rds",
                                package = "FishMap"
                                )

survey_data <- readr::read_rds(file = survey_data_file)

vmslogbook_data_file <- system.file("original_data",
                                "Solea_solea",
                                "vmslogbook_data.Rds",
                                package = "FishMap"
                                )

vmslogbook_data <- readr::read_rds(file = vmslogbook_data_file)

study_domain_file <- system.file("original_data",
                                "Solea_solea",
                                "study_domain.Rds",
                                package = "FishMap"
                                )

study_domain <- readr::read_rds(file = study_domain_file)


# prepare and load model inputs
fm_data_inputs <- fm_load_data(species = "Solea_solea",
                         fleet = c("OTB_DEF_>=70_0","OTB_CEP_>=70_0","OTT_DEF_>=70_0"),
                         fitted_data = "biomass",
                         survey_data = survey_data,
                         vmslogbook_data = vmslogbook_data,
                         study_domain = study_domain,
                         year_start = 2018,
                         year_end = 2018,
                         month_start = 11,
                         month_end = 11,
                         time_step = "Month",
                         k = 0.25,
                         grid_xmin = -6,
                         grid_xmax = 0,
                         grid_ymin = 42,
                         grid_ymax = 48)

# Fit the model
fm_model_results <- fm_fit_model(fm_data_inputs = fm_data_inputs,
                                 SE = 1,
                                 data_source = 1,
                                 data_obs = 2,
                                 samp_process = 0,
                                 b_constraint = 2,
                                 cov_samp_process = 0,
                                 biomass_temporal = 1,
                                 sampling_temporal = 0,
                                 lf_link = 0,
                                 ref_data = "com",
                                 EM = "est_b",
                                 month_ref = 1)

# Generate figure outputs
fm_generate_graphs(fm_model_results = fm_model_results)
```

## Code of Conduct

Please note that the FishMap project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# FishMap 0.0.1

* Each script is documented and tested.
* The README explains how to install a package
* The user vignette explains how to use it
* Implementation of two tested models
* A vignette explains how to maintain unit tests
* Removal of old scripts
* Data.frames are passed to functions
* Schematic explains the structure of the package


dev/r_scripts_orig/main.R
#----------------------------------------------
## Toy example to run the spatio-temporal model
#----------------------------------------------

## Packages
library(dplyr)
library(INLA)
library(ggplot2)
library(raster)
library(sf)
library(stringr)
library(tidyr)
library(TMB)
library(tictoc)

#----- Flat 1 : Data preparation ----

## Load data
#-----------
message("Running step 1 -loading data-")
tic("Step 1 -loading data-")

species <- "Solea_solea" # species of interest
fleet <- c("OTB_DEF_>=70_0","OTB_CEP_>=70_0","OTT_DEF_>=70_0") # Fleet to filter
# The first one will be taken as reference level in the model

data_folder <- here::here("inst","original_data",species)

fitted_data <- "biomass" # "biomass" "presabs"

# Scientific data
n_survey <- 1 # number of surveys
load(file.path(data_folder,"survey_data.Rdata")) # scientific observations
scientific_observation <- "CPUE" # 'CPUE' or 'Density'

survey_data_0 <- survey_data %>% ungroup %>% dplyr::select(-layer)

# 'VMS x logbook' data
load(file.path(data_folder,"vmslogbook_data.Rdata")) # 'vms x logbooks' observations

select_aggreg_level <- paste(fleet,collapse = "|")
vmslogbook_data <- vmslogbook_data %>%
  filter(str_detect(LE_MET_level6,select_aggreg_level))

vmslogbook_data$LE_MET_level6 <- factor(as.character(vmslogbook_data$LE_MET_level6),levels = fleet)

vmslogbook_data_0 <- vmslogbook_data %>% ungroup %>% dplyr::select(-layer)

## Time series
#-------------
year_start <- 2018
year_end <- 2018
year_vec <- year_start:year_end
month_start <- 10
month_end <- 12
month_vec <- month_start:month_end
time_step <- "Month" # Month or Quarter

if(time_step == "Month"){

  time.step_df <- expand.grid(month_vec,year_vec)
  colnames(time.step_df) <- c("Month","Year")

  time.step_df <- time.step_df %>%
    arrange(Year,Month) %>%
    mutate(Month = ifelse(Month < 10,paste0("0",Month),Month)) %>%
    mutate(Year_Month = paste0(Year,"_",Month)) %>%
    mutate(t = 1:nrow(time.step_df))
  time.step_df$Year <- as.character(time.step_df$Year)
  time.step_df$Month <- as.character(time.step_df$Month)

}else if(time_step == "Quarter"){

  time.step_df <- expand.grid(1:4,all_years)
  colnames(time.step_df) <- c("Quarter","Year")

  time.step_df <- time.step_df %>%
    arrange(Year,Quarter) %>%
    mutate(Quarter = ifelse(Quarter < 10,paste0("0",Quarter),Quarter)) %>%
    mutate(Year_Quarter = paste0(Year,"_",Quarter)) %>%
    mutate(t = 1:nrow(time.step_df))
  time.step_df$Year <- as.character(time.step_df$Year)
  time.step_df$Quarter <- as.character(time.step_df$Quarter)

}


## Configure spatial domain
#--------------------------
grid_xmin <- -6
grid_xmax <- 0
grid_ymin <- 42
grid_ymax <- 48
grid_limit <- extent(c(grid_xmin,grid_xmax,grid_ymin,grid_ymax))

grid_projection <- "+proj=longlat +datum=WGS84"

resol <- 0.05 # resolution of the discretization grid
create_mesh <- "from_shapefile"
# from_shapefile: the mesh will be more regular on the grid
# from_data: the mesh will be denser in the areas where there are data

# Mesh parameterization
# reduce the mesh size (k = 0.25 now) reduces the number of knots at which the spatial random effect is computed.
k <- 0.75
Alpha <- 2

load(file.path(data_folder,"study_domain.Rdata"))

## Load domain / mesh / spde object
source(here::here("inst","original_scripts","domain_mesh_spde.R"))

## Shape scientific data
source(here::here("inst","original_scripts","shape_sci_data_st.R"))

## Shape commercial data
source(here::here("inst","original_scripts","shape_vmslogbook_data_st.R"))

if(fitted_data=="presabs"){
  y_com_i[which(y_com_i > 0)] <- 1
  y_sci_i[which(y_sci_i > 0)] <- 1
  lf_link <- 1 # logit link
}


## Covariates
#------------
# load(file.path(data_folder,"bathy_pred.Rdata"))
bathy_pred = rep(0,nrow(loc_x))
cov_x_pred <- matrix(data = bathy_pred, ncol = 1)

# load(file.path(data_folder,"bathy_com.Rdata"))
bathy_com = rep(0,nrow(vmslogbook_data_2))
cov_x_com <- matrix(data = bathy_com, ncol = 1)

# load(file.path(data_folder,"bathy_sci.Rdata"))
bathy_sci = rep(0,nrow(survey_data_2))
cov_x_sci <- matrix(data = bathy_sci, ncol = 1)

# finished step 1 -loading data-
toc()

#---------- Flat 2 : Model init and fit ----

## Fit model
#-----------
message("Running step 2 -compile model-")
tic("Step 2 -compile model-")

## Model configuration
SE <- 1 # run ADREPORT - 0: no, 1: yes
data_source <- 1 # 1: integrated model (scientific + commercial data), 2: scientific model, 3: commercial model
data_obs <- 2 # observation model for biomass - 1: zero_inflated gamma, 2: zero-inflated lognormal, 3: lognormal
samp_process <- 0 # Sampling process - 0: no sampling process, 1: inhomogeneous Poisson point process
b_constraint <- 2 # put constraint on b parameters - 1: b are positive, 2: no constraints
const_spphab <- 1 # Species-habitat relationship - 1: constant in time
cov_samp_process <- 0 # covariate in the sampling process - 0: none, 1: covariate in the samplign process
biomass_temporal <- 1 # Account for temporal correlation in biomass - 0: no, 1: AR1
sampling_temporal <- 0 # Account for temporal correlation in sampling process - 0: no, 1: AR1
anisotropy <- 0 # Account for anisotropy
lf_link <- 0 # link function of the latent field - 0: log (biomass data), 1: logit (presence absence data)
ref_data <- "com" # reference data - "com": commercial (default) or "sci": scientific
EM <- "est_b" # "est_b": b is estimated, "fix_b": b is fixed
month_ref <- 1 # reference month (here January)
cov_samp_process <- 0 # 0: no covariate in the sampling process, 1: put covariate in the sampling process
compute_sd <- FALSE

xfb_x <- NULL # TO DELETE
weights_com <- 1 # TO DELETE

## Build Data, Params and Map objects for model fitting
source(here::here("inst","original_scripts","build_data_params_map.R"))

# Add link to path
if (.Platform$OS.type == "windows"){
  fixwinpath <- function(){
    PATH <- Sys.getenv("PATH")
    PATH <- paste0(R.home(), "/bin/x64;", PATH)
    PATH <- paste0("c:/Rtools/mingw64/bin;", PATH)
    Sys.setenv(PATH=PATH)
  }
  fixwinpath()
  shell("where g++")
  shell("where gdb")
  shell("where Rterm")
}

## Model compilation
TMB::compile(here::here("inst","model.cpp"),"-O1 -g",DLLFLAGS="")
dyn.load( dynlib(here::here("inst","model") ))

# finished step 2 -compile model-
toc()

## Fit model
message("Running step 3 -fit model-")
tic("Step 3 -fit model-")

# ## Debugging
# source("r/function/MakeADFun_windows_debug.R")
# MakeADFun_windows_debug(cpp_name = "inst/model",  data=Data, parameters=Params,  random=Random)
# TMB::gdbsource("inst/model.R",interactive = T) ## Non-interactive
# dyn.unload( dynlib( "inst/model" ) )
# #-----------

obj = MakeADFun( data=Data, parameters=Params,  random=Random, map = Map, silent = TRUE,hessian = TRUE )
# next line takes a long time ----
obj$fn( obj$par )

# Parameters boundary for optimization
Lower <- -50
Upper <- 50

if(T %in% str_detect(names(obj$par),"rho_")){ # constraints on the bounds of rho

  Lower <- rep(-50,length(obj$par))
  Upper <- rep(50,length(obj$par))

  Lower[which(str_detect(names(obj$par),"rho_"))] <- -0.99
  Upper[which(str_detect(names(obj$par),"rho_"))] <- 0.99

}

# next line takes a VERY long time ----
opt = nlminb( start=obj$par, objective=obj$fn, gradient=obj$gr, lower=Lower, upper=Upper, control=list(trace=1, maxit=200))
opt[["diagnostics"]] = data.frame( "Param"=names(obj$par), "Lower"=-Inf, "Est"=opt$par, "Upper"=Inf, "gradient"=obj$gr(opt$par) )

#save intermediate opt
if (Sys.getenv("FISHMAP_UPDATE_OUTPUTS") == "TRUE") {
  output_dir <- Sys.getenv("FISHMAP_OUTPUT_DIR")
  saveRDS(object = opt, file = file.path(output_dir, "opt_output.rds"))
  saveRDS(object = obj, file = file.path(output_dir, "obj_input.rds"))
}

report = obj$report() # output values
converge=opt$convergence # convergence test
# next line takes a VERY long time ----
if(compute_sd) SD = sdreport(obj,bias.correct=FALSE,ignore.parm.uncertainty=TRUE) # compute standard deviation

#save intermediate opt
if (Sys.getenv("FISHMAP_UPDATE_OUTPUTS") == "TRUE") {
  output_dir <- Sys.getenv("FISHMAP_OUTPUT_DIR")
  saveRDS(object = report, file = file.path(output_dir,"report_output.rds"))
  saveRDS(object = converge, file = file.path(output_dir, "converge_output.rds"))
}

dyn.unload( dynlib( here::here("inst","model" ) ))

# Finished step 3 - fit model-
toc()

#---- Flat 3 : Generate graphs -----
## Plot model outputs
#--------------------
message("Running step 4 -plot graphs-")
tic("Step 4 -plot graphs-")

if(nrow(time.step_df)>1){

  pred_df <- cbind(loc_x[,c("long","lati")],S_x=report$S_p[1:nrow(loc_x),]) %>%
    pivot_longer(cols = starts_with("S_x."),names_to = "t", values_to = "S_x") %>%
    mutate(t = as.numeric(str_replace(t,"S_x.",""))) %>%
    inner_join(time.step_df)

  pred_sf <- st_as_sf(pred_df,coords = c("long","lati"))

  pred_plot <- ggplot(pred_sf)+
    geom_sf(aes(col=S_x))+
    scale_color_distiller(palette = "Spectral")+
    facet_wrap(.~Year_Month)


}else if(nrow(time.step_df)==1){

  pred_df <- cbind(loc_x[,c("long","lati")],S_x=report$S_p[1:nrow(loc_x),])

  pred_sf <- st_as_sf(pred_df,coords = c("long","lati"))

  pred_plot <- ggplot(pred_sf)+
    geom_sf(aes(col=S_x))+
    scale_color_distiller(palette = "Spectral")

}

plot(pred_plot)

# plot eta if sampling is active
if (samp_process == 1){
  for(i in 1:3){
    eta_df <- cbind(loc_x[,c("long","lati")],eta=report$lambda_p[1:nrow(loc_x),,i]) %>%
      pivot_longer(cols = starts_with("eta."),names_to = "t", values_to = "eta") %>%
      mutate(t = as.numeric(str_replace(t,"eta.",""))) %>%
      inner_join(time.step_df)

    eta_sf <- st_as_sf(eta_df,coords = c("long","lati"))

    eta_plot <- ggplot(eta_sf)+
      geom_sf(aes(col=log(eta)))+
      scale_color_distiller(palette = "Spectral")+
      facet_wrap(.~Year_Month)

    plot(eta_plot)
  }
}

# finished step 4 -plot graph-
toc()
